src/storage/src/filter_key_extractor.rs

```diff
@@ -376,11 +375,52 @@ impl FilterKeyExtractorManager {
     /// Acquire a `MultiFilterKeyExtractor` by `table_id_set`
     /// Internally, try to get all `filter_key_extractor` from `hashmap`. Will block the caller if
     /// `table_id` does not util version update (notify), and retry to get
+    async fn acquire(&self, table_id_set: HashSet<u32>) -> HummockResult<FilterKeyExtractorImpl> {
+        self.inner.acquire(table_id_set).await
+    }
+}
+
+#[derive(Clone)]
+pub enum FilterKeyExtractorManagerFactory {
+    FilterKeyExtractorManagerRef(Arc<FilterKeyExtractorManager>),
+    ServerlessFilterKeyExtractorManager(FilterKeyExtractorBuilder),
+}
+
+impl FilterKeyExtractorManagerFactory {
     pub async fn acquire(
         &self,
         table_id_set: HashSet<u32>,
     ) -> HummockResult<FilterKeyExtractorImpl> {
-        self.inner.acquire(table_id_set).await
+        match self {
+            FilterKeyExtractorManagerFactory::FilterKeyExtractorManagerRef(
+                filter_key_exactor_manager,
+            ) => filter_key_exactor_manager.acquire(table_id_set).await,
+            FilterKeyExtractorManagerFactory::ServerlessFilterKeyExtractorManager(
+                filter_key_extractor_builder,
+            ) => filter_key_extractor_builder.acquire(table_id_set),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct FilterKeyExtractorBuilder {
+    id_to_table: HashMap<u32, Table>,
+}
+
+impl FilterKeyExtractorBuilder {
+    pub fn new(id_to_table: HashMap<u32, Table>) -> Self {
+        Self { id_to_table }
+    }
+
+    fn acquire(&self, table_id_set: HashSet<u32>) -> HummockResult<FilterKeyExtractorImpl> {
+        let mut multi_filter_key_extractor = MultiFilterKeyExtractor::default();
+        for table_id in table_id_set {
+            if let Some(table) = self.id_to_table.get(&table_id) {
```

Comment:
What happen if some tables in the `table_id_set` are absent in `id_to_table`? Should we return an error for that?

Reply:
Let's return error as a backstop, and proxy should guarantee that this will not happen.